From mtosatti@redhat.com  Thu Sep  3 14:39:10 2009
From: Avi Kivity <avi@redhat.com>
Date: Thu,  6 Aug 2009 14:40:02 -0300
Subject: KVM: Make paravirt tlb flush also reload the PAE PDPTRs
To: stable@kernel.org
Cc: Avi Kivity <avi@redhat.com>
Message-ID: <1249580407-21883-23-git-send-email-mtosatti@redhat.com>


From: Avi Kivity <avi@redhat.com>

(cherry picked from commit a8cd0244e9cebcf9b358d24c7e7410062f3665cb)

The paravirt tlb flush may be used not only to flush TLBs, but also
to reload the four page-directory-pointer-table entries, as it is used
as a replacement for reloading CR3.  Change the code to do the entire
CR3 reloading dance instead of simply flushing the TLB.

Signed-off-by: Avi Kivity <avi@redhat.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
---
 arch/x86/kvm/mmu.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/arch/x86/kvm/mmu.c
+++ b/arch/x86/kvm/mmu.c
@@ -2233,7 +2233,7 @@ static int kvm_pv_mmu_write(struct kvm_v
 
 static int kvm_pv_mmu_flush_tlb(struct kvm_vcpu *vcpu)
 {
-	kvm_x86_ops->tlb_flush(vcpu);
+	kvm_set_cr3(vcpu, vcpu->arch.cr3);
 	return 1;
 }
 
